### PR TITLE
Add support for template arguments

### PR DESCRIPTION
**Pull request by prashantv**
Template arguments are of the format ${argname:fallback}. Arguments can
be specified on the commandline using `--arg` or `-A`. Arguments are
parsed as YAML values, and allow more complex types (e.g., lists) to be
specified as arguments.

We only use template arguments for request bodies currently.

Argument parsing is based on @abhinav's work for YARPC configuration.

### CODE REVIEW

File: templateargs/argparser/arg.rl [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    write data;
+}%%
+
+// Var is the parsed envvar string.
+type Var struct {
+  Name  string
+  Value string
+}
+
+// Parse parses strings in the format, ${NAME} and ${NAME:VALUE}.
```
Comment (kriskowal) [inline templateargs/argparser/arg.rl]: Add a Lit struct and return LitOrVar. Lit should produce a string with `\$ => $` and `\\ => \`.

File: templateargs/process.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+		if err := yaml.Unmarshal([]byte(v), &unmarshalled); err != nil {
+			return nil, fmt.Errorf("failed to process %q with value %q: %v", k, v, err)
+		}
+		argsUnmarshalled[k] = unmarshalled
+	}
+
+	return processMap(req, argsUnmarshalled)
+}
+
+func processString(v string, args map[string]interface{}) (interface{}, error) {
```
Comment (kriskowal) [inline templateargs/process.go]: This function should unconditionally parse the string in its DSL, including escaping `\$` and `\\`. The grammar should return a simple AST with either a Lit or a Var term, and this function should just switch on that.

This leaves open the possibility of extending the grammar to support string interpolation like `${host}:${port}`. We can forbid this for now.

File: testdata/templates/bad-arg.yaml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+service: foo
+request:
+  name: ${user:foo
```
Comment (kriskowal) [inline testdata/templates/bad-arg.yaml]: Needs test for `\${user}`. Would be good to test `\\`, `\$`.

Comment (kriskowal) [inline testdata/templates/bad-arg.yaml]: Might be good to test requests with embedded non-string arguments, e.g., as in `yab -A peer-list:$(cat peers.yaml)`

File: templateargs/process.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	// If a string starts with \, avoid processing the value and strip the prefix.
+	if strings.HasPrefix(v, `\`) {
+		return v[1:], nil
+	}
+
+	// Anything that starts with "${ " should be processed as a template arg.
+	if !strings.HasPrefix(v, "${") {
```
Comment (abhinav) [inline templateargs/process.go]: and ends with `}`?

File: testdata/templates/bad-arg.yaml (hunk 3 above)
Comment (prashantv) [inline testdata/templates/bad-arg.yaml]: We cover a lot of the edge cases such as `\\` and `\$` in `TestProcessString` and `TestProcessMap` already. but added a couple more to `args.yaml` -- it has a fallback which is not a string, a list replacement, and testing for escapes.

File: testdata/templates/args.yaml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+service: foo
+headers:
+  header1: ${user:foo}
```
Comment (prashantv) [inline testdata/templates/args.yaml]: Note: this is not replaced at the moment, template arguments are used for request bodies.

Before 1.0, we may want to think about processing everything in the YAML file? We would probably decode the template to a `map`, use `templateargs.ProcessMap`, then pass it through `mapdecode`. Thoughts @kriskowal @abhinav

Comment (abhinav) [inline testdata/templates/args.yaml]: Sounds reasonable. Although `mapdecode` expects to decode into a known struct shape, not dynamic Thrift shapes.

Comment (prashantv) [inline testdata/templates/args.yaml]: Yep, we want to use it to load into the `Template` struct (known Go shape). Won't affect the body which is still a map.

File: options.go [hunk 6]
```diff
@@ -52,6 +52,7 @@ type RequestOptions struct {
 	Health       bool              `long:"health" description:"Hit the health endpoint, Meta::health"`
 	Timeout      timeMillisFlag    `long:"timeout" default:"1s" description:"The timeout for each request. E.g., 100ms, 0.5s, 1s. If no unit is specified, milliseconds are assumed."`
 	YamlTemplate string            `short:"y" long:"yaml-template" description:"Send a tchannel request specified by a yaml template"`
+	TemplateArgs map[string]string `short:"A" long:"arg" description:"A list of key-value template arguments"`
```
Comment (prashantv) [inline options.go]: 👍

**Pull request merged**